![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2FMLOps%2FModel+Evaluation&file=model-evaluation-classification-multi-label.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2FMLOps%2FModel%2520Evaluation%2Fmodel-evaluation-classification-multi-label.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/MLOps/Model%20Evaluation/model-evaluation-classification-multi-label.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
</table>

# Evaluations For Multi-Label Classification Models

This workflow, part of our [MLOps](../readme.md) series that dives into [Model Evaluation](./readme.md), provides a comprehensive, end-to-end example of building and evaluating a machine learning model directly within Google Cloud's Vertex AI ecosystem.

Specifically, this guide will walk you through how to:

* **Prepare Data at Scale:** Begin by exploring and efficiently preparing a large public dataset sourced from **BigQuery**, demonstrating how to leverage both powerful **SQL** queries and the interactive **BigFrames API** (pandas interface) for machine learning data readiness.
* **Build Robust Models with Scikit-learn:** Construct a complete **Scikit-learn model pipeline** that seamlessly integrates additional feature engineering steps with the training of a classification model.
* **Register Models in Vertex AI:** Master the crucial MLOps step of saving your trained model and **registering it as a version within the Vertex AI Model Registry** for centralized management and version control.
* **Generate & Upload Custom Metrics:** Delve into preparing **custom, detailed evaluation metrics** using **Scikit-learn**, and then strategically **load these results to your versioned model in the Vertex AI Model Registry**. This includes showcasing how to generate and upload metrics for specific **data slices** to gain granular performance insights.
* **Review & Retrieve Evaluations:** Conclude by learning how to easily **review and programmatically retrieve these comprehensive evaluation results** directly from the Model Registry UI and via the **Vertex AI SDK**, empowering you to effectively track and compare model performance over time.
* **Generate and Work With Slices** Create metrics for slices of the data, in this case the class levels of the multi-class model.  Learn to generate and upload these to model evaluations in the Vertex AI Model Registry.

**The Model:**

This workflow uses a data source of Stack Overflow posts.  It creates a predictive model that predicts which tags apply to a post (e.g., 'python', 'pandas', 'data-science') - a multi-label classifier.

---
## Colab Setup

To run this notebook in Colab run the cells in this section.  Otherwise, skip this section.

This cell will authenticate to GCP (follow prompts in the popup).

In [1]:
PROJECT_ID = 'statmike-mlops-349915' # replace with project ID

In [2]:
try:
    import google.colab
    from google.colab import auth
    auth.authenticate_user()
    !gcloud config set project {PROJECT_ID}
except Exception:
    pass

---
## Installs

The list `packages` contains tuples of package import names and install names.  If the import name is not found then the install name is used to install quitely for the current user.

In [4]:
# tuples of (import name, install name, min_version)
packages = [
    ('bigframes', 'bigframes'),
    ('sklearn', 'scikit-learn'),
    ('numpy', 'numpy'),
    ('shap', 'shap'),
    ('google.cloud.aiplatform', 'google-cloud-aiplatform'), 
    ('google.cloud.storage', 'google-cloud-storage'),
    ('google.cloud.bigquery', 'google-cloud-bigquery'),
    ('bigquery_magics', 'bigquery-magics'),
    ('matplotlib', 'matplotlib'),
    ('pandas', 'pandas')
]

import importlib
install = False
for package in packages:
    if not importlib.util.find_spec(package[0]):
        print(f'installing package {package[1]}')
        install = True
        !pip install {package[1]} -U -q --user
    elif len(package) == 3:
        if importlib.metadata.version(package[0]) < package[2]:
            print(f'updating package {package[1]}')
            install = True
            !pip install {package[1]} -U -q --user

### API Enablement

In [5]:
!gcloud services enable aiplatform.googleapis.com

### Restart Kernel (If Installs Occured)

After a kernel restart the code submission can start with the next cell after this one.

In [6]:
if install:
    import IPython
    app = IPython.Application.instance()
    app.kernel.do_shutdown(True)
    IPython.display.display(IPython.display.Markdown("""<div class=\"alert alert-block alert-warning\">
        <b>⚠️ The kernel is going to restart. Please wait until it is finished before continuing to the next step. The previous cells do not need to be run again⚠️</b>
        </div>"""))

---
## Setup

inputs:

In [7]:
project = !gcloud config get-value project
PROJECT_ID = project[0]
PROJECT_ID

'statmike-mlops-349915'

In [8]:
REGION = 'us-central1'
SERIES = 'mlops'
EXPERIMENT = 'evaluation-classification-multi-label'

# Set the name of GCS Bucket to read/write to
GCS_BUCKET = PROJECT_ID

# Data source for this workflow
BQ_SOURCE = 'bigquery-public-data.stackoverflow.posts_questions'

# make this the BigQuery Project / Dataset / Table prefix to store results
BQ_PROJECT = PROJECT_ID
BQ_DATASET = SERIES.replace('-', '_')
BQ_TABLE = EXPERIMENT
BQ_REGION = REGION[0:2] # use a multi region

packages:

In [10]:
import os
import joblib
import sklearn.metrics
import sklearn.ensemble
import sklearn.pipeline
import sklearn.compose
import sklearn.preprocessing
import shap
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import bigframes.pandas as bpd
from google.cloud import aiplatform
from google.cloud import bigquery
from google.cloud import storage

clients:

In [11]:
# vertex ai clients
aiplatform.init(project = PROJECT_ID, location = REGION, experiment = SERIES+'-'+EXPERIMENT)

# gcs storage client
gcs = storage.Client(project = GCS_BUCKET)
bucket = gcs.bucket(GCS_BUCKET)

# bigquery client
bq = bigquery.Client(project = PROJECT_ID)

# bigframes setup
bpd.options.bigquery.project = PROJECT_ID

# bigquery cell magics load
%load_ext bigquery_magics

Parameters:

In [12]:
DIR = f"files/{EXPERIMENT}"

Environment:

In [13]:
if not os.path.exists(DIR):
    os.makedirs(DIR)

---
## Review Data Source: Stack Overflow Posts

This project uses a sample of data from the following data source as a tutorial.  

The source table is a BigQuery Public Dataset table of Stack Overflow posted questions and contains the title and body text as well as associated tags for the posts.  

There are 23,020,127 observations including 2,116,212 from the year 2017 that are processed for use in this workflow.

### Review BigQuery table:

Use the [BigQuery BigFrames](https://cloud.google.com/bigquery/docs/use-bigquery-dataframes) package, setup above, to treat BigQuery tables like dataframes.  This has the advantage of doing the computing within BigQuery rather than pulling the data locally.

In [14]:
source_data = bpd.read_gbq(BQ_SOURCE)
source_data.head()

id                                              title  \
0  70258907  Issue when using host file in docker container...   
1  60254980  I have click event attached to elements <li>, ...   
2  47012267           Make a better grid of points with python   
3  67073082  Jetpack Compose beta cannot go edge-to-edge du...   
4   5520689                     Analyzing Recursive Algorithms   

                                                body  accepted_answer_id  \
0  <p>My issue is the &quot;auto-encryption&quot;...            70422908   
1  <p>I have click event attached to elements <co...                <NA>   
2  <p>I can make simple grid of points in a recta...                <NA>   
3  <p>I have upgraded Compose for my app from <co...                <NA>   
4  <p>I've often been slightly stumped by recursi...             5657512   

   answer_count  comment_count community_owned_date  \
0             1              0                 <NA>   
1             2              2                 <NA>   
2             1              2                 <NA>   
3             1              5                 <NA>   
4             2              4                 <NA>   

                      creation_date  favorite_count  \
0  2021-12-07 10:56:44.280000+00:00            <NA>   
1  2020-02-17 01:53:31.527000+00:00            <NA>   
2  2017-10-30 09:51:57.940000+00:00               1   
3  2021-04-13 10:26:27.813000+00:00            <NA>   
4  2011-04-02 02:19:08.940000+00:00               1   

                 last_activity_date                    last_edit_date  \
0  2021-12-20 14:38:06.993000+00:00  2021-12-07 12:44:57.153000+00:00   
1  2020-02-17 08:31:50.267000+00:00  2020-02-17 08:31:50.267000+00:00   
2  2017-10-30 10:19:21.197000+00:00                              <NA>   
3  2021-07-13 11:18:35.677000+00:00  2021-04-13 12:25:11.390000+00:00   
4  2011-04-14 01:14:42.780000+00:00  2011-04-02 02:58:21.543000+00:00   

  last_editor_display_name  last_editor_user_id owner_display_name  \
0                     <NA>               835636               <NA>   
1                     <NA>              4370109               <NA>   
2                     <NA>                 <NA>               <NA>   
3                     <NA>               975887               <NA>   
4                     <NA>                33833               <NA>   

   owner_user_id parent_id  post_type_id  score  \
0         835636      <NA>             1      0   
1        3230529      <NA>             1     -2   
2        8855484      <NA>             1      1   
3         975887      <NA>             1      0   
4          33833      <NA>             1      1   

                                                tags  view_count  
0            docker|apache-karaf|apache-felix|jasypt         126  
1                    javascript|jquery|jquery-events         659  
2                                numpy|grid|boundary        5159  
3  android|android-jetpack-compose|android-immersive        1280  
4                                 algorithm|analysis        1232  

[5 rows x 20 columns]

In [15]:
source_data.info()

<class 'bigframes.dataframe.DataFrame'>
Index: 23020127 entries, 0 to 23020126
Data columns (total 20 columns):
  #  Column                    Dtype
---  ------------------------  ------------------------------
  0  id                        Int64
  1  title                     string
  2  body                      string
  3  accepted_answer_id        Int64
  4  answer_count              Int64
  5  comment_count             Int64
  6  community_owned_date      timestamp[us, tz=UTC][pyarrow]
  7  creation_date             timestamp[us, tz=UTC][pyarrow]
  8  favorite_count            Int64
  9  last_activity_date        timestamp[us, tz=UTC][pyarrow]
 10  last_edit_date            timestamp[us, tz=UTC][pyarrow]
 11  last_editor_display_name  string
 12  last_editor_user_id       Int64
 13  owner_display_name        string
 14  owner_user_id             Int64
 15  parent_id                 string
 16  post_type_id              Int64
 17  score                     Int64
 18  tags         

In [16]:
source_data.describe(include = 'all')

id     title      body  accepted_answer_id  \
count         23020127.0  23020127  23020127          11755280.0   
nunique             <NA>  23006006  23017995                <NA>   
mean     38214250.492806      <NA>      <NA>     34495650.183957   
std      21099662.230776      <NA>      <NA>     21071281.966936   
min                  4.0      <NA>      <NA>                 7.0   
25%           20082502.0      <NA>      <NA>          15560941.0   
50%           38244069.0      <NA>      <NA>          32850998.0   
75%           56483609.0      <NA>      <NA>          52647556.0   
max           73842327.0      <NA>      <NA>          73842204.0   

         answer_count  comment_count  community_owned_date  creation_date  \
count      23020127.0     23020127.0                 11355       23020127   
nunique          <NA>           <NA>                  <NA>           <NA>   
mean         1.478016       1.985216                  <NA>           <NA>   
std          1.453571       2.662199                  <NA>           <NA>   
min               0.0            0.0                  <NA>           <NA>   
25%               1.0            0.0                  <NA>           <NA>   
50%               1.0            1.0                  <NA>           <NA>   
75%               2.0            3.0                  <NA>           <NA>   
max             518.0          108.0                  <NA>           <NA>   

         favorite_count  last_activity_date  last_edit_date  \
count         5029531.0            23020127        12559922   
nunique            <NA>                <NA>            <NA>   
mean           2.824621                <NA>            <NA>   
std           21.154099                <NA>            <NA>   
min                 0.0                <NA>            <NA>   
25%                 1.0                <NA>            <NA>   
50%                 1.0                <NA>            <NA>   
75%                 2.0                <NA>            <NA>   
max             11784.0                <NA>            <NA>   

         last_editor_display_name  last_editor_user_id  owner_display_name  \
count                      265452           12361577.0              563534   
nunique                     53951                 <NA>              130869   
mean                         <NA>       3487890.377558                <NA>   
std                          <NA>       4029319.918362                <NA>   
min                          <NA>                 -1.0                <NA>   
25%                          <NA>             472792.0                <NA>   
50%                          <NA>            1839439.0                <NA>   
75%                          <NA>            5050401.0                <NA>   
max                          <NA>           20080937.0                <NA>   

          owner_user_id  parent_id  post_type_id       score      tags  \
count        22562265.0          0    23020127.0  23020127.0  23020127   
nunique            <NA>          0          <NA>        <NA>   8448317   
mean     4898520.717387       <NA>           1.0    2.214833      <NA>   
std       4691874.94353       <NA>           0.0   25.754884      <NA>   
min                 1.0       <NA>           1.0      -146.0      <NA>   
25%           1149526.0       <NA>           1.0         0.0      <NA>   
50%           3276418.0       <NA>           1.0         0.0      <NA>   
75%           7347345.0       <NA>           1.0         2.0      <NA>   
max          20081043.0       <NA>           1.0     26621.0      <NA>   

           view_count  
count      23020127.0  
nunique          <NA>  
mean      2827.689736  
std      24555.661991  
min               1.0  
25%              97.0  
50%             348.0  
75%            1247.0  
max        11649204.0  

[9 rows x 20 columns]

### Understand The Data Source With `ML.DESCRIBE_DATA`

Reviewing a few records, like above, gives a good sense of how the data is arranged. Before proceeding with machine learning techniques it is important to understand more about these raw columns.  Are they ready to use a features in a model or is some form of feature engineering needed first?  For this, the distribution of values is an important starting point.  

While SQL could be used to look at the distribution, it would be a time consuming process and requires different techniques for different data types like numerical, string, boolean, dates, times, array and struct version of these, and arrays of structs.

To make this process fast and simple, the new [`ML.DESCRIBE_DATA`](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-describe-data) function is used to get a single row for each column the describes the data distribution:
- `top_k`: get the top 3 most frequent categories for string columns (default = 1)
- `num_quantiles`: get 4 quantiles for numerical columns (default = 2)

In [17]:
%%bigquery
SELECT *
FROM ML.DESCRIBE_DATA(
    TABLE `bigquery-public-data.stackoverflow.posts_questions`,
    STRUCT(3 as top_k, 4 as num_quantiles)
)

Query is running:   0%|          |

Downloading:   0%|          |

name  num_rows  num_nulls  num_zeros  \
0         accepted_answer_id  23020127   11264847          0   
1               answer_count  23020127          0    3316808   
2                       body  23020127          0       <NA>   
3              comment_count  23020127          0    9363819   
4       community_owned_date  23020127   23008772       <NA>   
5              creation_date  23020127          0       <NA>   
6             favorite_count  23020127   17990596     725181   
7                         id  23020127          0          0   
8         last_activity_date  23020127          0       <NA>   
9             last_edit_date  23020127   10460205       <NA>   
10  last_editor_display_name  23020127   22754675       <NA>   
11       last_editor_user_id  23020127   10658550          0   
12        owner_display_name  23020127   22456593       <NA>   
13             owner_user_id  23020127     457862          0   
14                 parent_id  23020127   23020127       <NA>   
15              post_type_id  23020127          0          0   
16                     score  23020127          0   10525850   
17                      tags  23020127          0       <NA>   
18                     title  23020127          0       <NA>   
19                view_count  23020127          0          0   

                                                  min  \
0                                                   7   
1                                                   0   
2   " runat="server" />\n    \n\n<p>But this code ...   
3                                                   0   
4                          2008-08-12 04:59:35.017+00   
5                          2008-07-31 21:42:52.667+00   
6                                                   0   
7                                                   4   
8                          2008-09-04 12:50:25.060+00   
9                          2008-08-03 21:38:52.623+00   
10                                        &#193;lvaro   
11                                                 -1   
12                                        &#193;lvaro   
13                                                  1   
14                                               None   
15                                                  1   
16                                               -146   
17                                                 .a   
18  ! ( NULL || 0 || '' ) if condition in PHP at t...   
19                                                  1   

                                                  max          mean  \
0                                            73842204  3.449565e+07   
1                                                 518  1.478016e+00   
2   مرحبا بك\n    {{authService.decodedToken?.uniq...           NaN   
3                                                 108  1.985216e+00   
4                          2022-09-19 13:19:19.410+00           NaN   
5                          2022-09-25 05:56:32.863+00           NaN   
6                                               11784  2.824621e+00   
7                                            73842327  3.821425e+07   
8                          2022-09-25 05:56:36.210+00           NaN   
9                          2022-09-25 05:56:30.490+00           NaN   
10                                                Мסž           NaN   
11                                           20080937  3.487890e+06   
12                                                홍한석           NaN   
13                                           20081043  4.898521e+06   
14                                               None           NaN   
15                                                  1  1.000000e+00   
16                                              26621  2.214833e+00   
17                                           zynq|soc           NaN   
18  � symbols in whole website from MySQL retrieve...           NaN   
19                                           11649204  2.8276

Some observations:
- The columns `body`, `title`, `tags` and `id` are always present, this is a great starting poing for a model to predict tags based on the contents of a post.

---
## Prepare Data Source

The data preparation includes adding splits for machine learning with a column named `splits` with values for training (`TRAIN`), validation (`VALIDATE`), and for testing (`TEST`). 

>These steps could be done locally at training but are instead done in the source system, BigQuery in this case, which provides several advantages:
>
>-   **Single Source of Truth:** A single data preparation can benefit multiple model training jobs for different architectures or even different team members working on the same model. This ensures consistency and avoids duplication of effort.
>-   **Leverage BigQuery's Power:** BigQuery is highly optimized for large-scale data processing. Performing these operations directly in BigQuery leverages its distributed processing capabilities, making the preparation significantly faster and more efficient than local processing, especially for massive datasets.
>-   **Reduced Data Movement:** Preparing the data in BigQuery reduces the amount of data that needs to be moved out of BigQuery and into the training environment. This minimizes latency and potential bottlenecks associated with data transfer.
>-   **Data Versioning and Reproducibility:** By preparing the splits and unique ID in BigQuery, the specific dataset used for training can be easily tracked and versioned. This enhances the reproducibility of experiments and makes it easier to understand the provenance of the data used in a particular model.
>-   **Simplified Training Pipeline:** The training pipeline becomes simpler because it can directly read pre-split data from BigQuery, eliminating the need for complex splitting logic within the training code.
>-   **Pre-calculated Joins and Features:** BigQuery can be used to pre-calculate joins and engineer new features that are beneficial for the model. This can improve model performance and further reduce the workload during the training phase.
>
>**Further Considerations:**
>
>-   **Data Governance and Security:** BigQuery offers robust data governance and security features. Performing data preparation within BigQuery allows you to maintain control over access and ensure data quality.
>-   **Scalability:** This approach is highly scalable. As your dataset grows, BigQuery can handle the increased workload without requiring significant changes to your data preparation pipeline.
>-   **Cost Optimization:** While moving large amounts of data out of BigQuery can incur costs, performing the preparation steps within BigQuery and only extracting the necessary data for training can often be more cost-effective.
>
>By preparing the data in BigQuery, you create a streamlined, efficient, and reproducible workflow (pipeline) that leverages the strengths of the platform and sets your machine learning models up for success.


### Create/Recall Dataset

In [18]:
dataset = bigquery.Dataset(f"{BQ_PROJECT}.{BQ_DATASET}")
dataset.location = BQ_REGION
bq_dataset = bq.create_dataset(dataset, exists_ok = True)

### Create/Recall Table With Preparation For ML

Copy the data from the source while filtering rows, processing columns:
- Filter to a single years records: 2017
- `splits` column to randomly assign rows to 'TRAIN", "VALIDATE" and "TEST" groups
- Subset columns to `body`, `title`, and `id`
- **NOTE:** This sample the data further down by only assigning splits to about 10% of the data.  This is strickly for local testing in this example.

In [21]:
job = bq.query(f"""
CREATE OR REPLACE TABLE `{BQ_PROJECT}.{BQ_DATASET}.{BQ_TABLE}` AS
WITH
    base_data AS (
        SELECT
            id,                                -- Unique identifier for the question
            title,                             -- Raw title text
            body,                              -- Raw body text
            tags,                               -- Multi-label target (pipe-separated string)
            ROW_NUMBER() OVER (ORDER BY RAND()) as rn -- simple row numbering for splitting
        FROM
            `{BQ_SOURCE}` -- Source table
        WHERE
            title IS NOT NULL AND body IS NOT NULL AND tags IS NOT NULL
            AND LENGTH(TRIM(title)) > 0 AND LENGTH(TRIM(body)) > 0 AND LENGTH(TRIM(tags)) > 0
            -- Filter out deleted or non-question posts if necessary (e.g., post_type_id = 1 for questions)
            -- For simplicity, we'll keep it minimal here.
            AND creation_date BETWEEN '2017-01-01 00:00:00+00:00' AND '2017-12-31 23:59:59+00:00' -- Focus on a specific year for manageability
    )
SELECT
    * EXCEPT(rn),
    CASE
        WHEN rn <= 0.08 * COUNT(*) OVER () THEN 'TRAIN'
        WHEN rn <= 0.09 * COUNT(*) OVER () THEN 'VALIDATE'
        WHEN rn <= 0.10 * COUNT(*) OVER () THEN 'TEST'
        ELSE Null -- records not sampled for using in training here
    END AS splits
FROM
    base_data
""")
job.result()
(job.ended-job.started).total_seconds()

13.147

In [22]:
source_data_prepared = bpd.read_gbq(f"{BQ_PROJECT}.{BQ_DATASET}.{BQ_TABLE}", use_cache = False)
source_data_prepared.head()

id                                              title  \
0  45949298  When running my function my if statement is no...   
1  41810643  Finding a gps locations near another one based...   
2  44769160  Spring - Provide parsed JWT as Resource method...   
3  42025994       Scala slick retrieve data tables in parallel   
4  42289360  What is the correct way to test multiple andro...   

                                                body  \
0  <p>I am trying to make a simple bubble sort, a...   
1  <p>I have a<code>GpsLocation</code> model, thi...   
2  <p>In Spring framework, is there a way to pars...   
3  <p>I need to read data from two different tabl...   
4  <p>I have a working "client-server" program , ...   

                                                tags splits  
0  python|arrays|python-3.x|if-statement|bubble-sort   <NA>  
1            ruby-on-rails|ruby-on-rails-4|geocoding   <NA>  
2                                         spring|jwt   <NA>  
3                    multithreading|scala|jdbc|slick  TRAIN  
4  android|android-studio|android-emulator|emulation   <NA>  

[5 rows x 5 columns]

### Review the number of records for each of the data splits:

In [23]:
bq.query(f"""
SELECT splits,
    count(*) as count,
    ROUND(
        COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2
    ) AS percentage
FROM `{BQ_PROJECT}.{BQ_DATASET}.{BQ_TABLE}`
WHERE splits IS NOT NULL
GROUP BY splits
ORDER BY splits
""").to_dataframe()

splits   count  percentage
0      TEST   21162        10.0
1     TRAIN  169296        80.0
2  VALIDATE   21163        10.0

### Further Data Preparation



#### Rows Assigned To A Split (TRAIN, TEST, VALIDATE)

Avoid the unsampled rows where `splits = Null`:

In [24]:
# only use rows assigned to a split:
source_data_prepared_filtered = source_data_prepared[source_data_prepared['splits'].notna()]

#### Group Columns By Type

In [25]:
# Feature Columns
string_feature_cols = [
    'title',
    'body'
]

target_col = 'tags'
split_col = 'splits'
id_col = 'id' # Keep id for later lookup if needed, but not as a feature

#### Check For Columns With NaNs

In [26]:
source_data_prepared_filtered[string_feature_cols + [target_col]].isna().sum()

title    0
body     0
tags     0
dtype: Int64

#### Setup Pointers For Training

In [27]:
X_frame = source_data_prepared_filtered[string_feature_cols]
y_frame = source_data_prepared_filtered[target_col]
splits_frame = source_data_prepared_filtered[split_col]

print(f"There are {X_frame.shape[0]} training rows for {X_frame.shape[1]} raw features.")
print(f"Confirming the rows for splits and the label are {splits_frame.shape[0]} and {y_frame.shape[0]} respectively.")

There are 211621 training rows for 2 raw features.
Confirming the rows for splits and the label are 211621 and 211621 respectively.


---
## Train With Scikit-Learn

### Local Dataframes

Convert the Bigframes pointers to Pandas dataframes for the training objects:

In [28]:
X_train = X_frame[splits_frame == 'TRAIN'].to_pandas().reset_index(drop=True)
X_val = X_frame[splits_frame == 'VALIDATE'].to_pandas().reset_index(drop=True)
X_test = X_frame[splits_frame == 'TEST'].to_pandas().reset_index(drop=True)

y_train_raw = y_frame[splits_frame == 'TRAIN'].to_pandas().reset_index(drop=True)
y_val_raw = y_frame[splits_frame == 'VALIDATE'].to_pandas().reset_index(drop=True)
y_test_raw = y_frame[splits_frame == 'TEST'].to_pandas().reset_index(drop=True)

In [29]:
X_train.shape, y_train_raw.shape

((169296, 2), (169296,))

### Prepare Multi-Label Target

The target data need to be turned into a sparse matrics where columns represent individual tags and the values represent the presence (1) or absense (0) of the tag for the sample (row).

The `y_raw_train` data will be used to make a list of the most common tags and then train a `MultiLabelBinarizer` to create the sparse matrix.  This learned representation will then be applied to the test and validate data splits.

In [30]:
# Convert pipe-delimited strings into lists of tags for each split
y_train_raw_lists = y_train_raw.apply(lambda x: x.split('|') if pd.notna(x) and x.strip() != '' else [])
y_val_raw_lists = y_val_raw.apply(lambda x: x.split('|') if pd.notna(x) and x.strip() != '' else [])
y_test_raw_lists = y_test_raw.apply(lambda x: x.split('|') if pd.notna(x) and x.strip() != '' else [])


# Count tag frequency from TRAINING DATA ONLY to identify frequent tags
all_tags_train = [tag for sublist in y_train_raw_lists for tag in sublist]
tag_counts_train = pd.Series(all_tags_train).value_counts()

# Identify most frequent tags based on training data (e.g., frequency > 500)
frequent_tags_train = tag_counts_train[tag_counts_train >= 500].index.tolist()

# Filter the tag lists for all splits to only include frequent tags learned from training
y_train_raw_lists_filtered = y_train_raw_lists.apply(lambda tags_list: [tag for tag in tags_list if tag in frequent_tags_train])
y_val_raw_lists_filtered = y_val_raw_lists.apply(lambda tags_list: [tag for tag in tags_list if tag in frequent_tags_train])
y_test_raw_lists_filtered = y_test_raw_lists.apply(lambda tags_list: [tag for tag in tags_list if tag in frequent_tags_train])

In [31]:
# Initialize MultiLabelBinarizer to output sparse matrix
mlb = sklearn.preprocessing.MultiLabelBinarizer(sparse_output=True)

# Fit mlb ONLY on training data, then transform all splits
y_train_processed = mlb.fit_transform(y_train_raw_lists_filtered)
y_val_processed = mlb.transform(y_val_raw_lists_filtered)
y_test_processed = mlb.transform(y_test_raw_lists_filtered)

In [32]:
y_train_processed.shape

(169296, 127)

In [33]:
mlb.classes_.shape

(127,)

In [34]:
pd.DataFrame(y_train_processed[:5].toarray(), columns=mlb.classes_).head()

.net  ajax  algorithm  amazon-web-services  android  android-studio  \
0     0     0          0                    0        0               0   
1     0     0          0                    0        0               0   
2     0     0          0                    0        0               0   
3     0     0          0                    0        1               0   
4     0     0          0                    0        0               0   

   angular  angularjs  apache  apache-spark  ...  visual-studio  vue.js  \
0        0          0       0             0  ...              0       0   
1        0          0       0             0  ...              0       0   
2        0          0       0             0  ...              0       0   
3        0          0       0             0  ...              0       0   
4        0          0       0             0  ...              0       0   

   webpack  windows  winforms  wordpress  wpf  xamarin  xcode  xml  
0        0        0         0          0    0        0      0    0  
1        0        0         0          0    0        0      0    0  
2        0        0         0          0    0        0      0    0  
3        0        0         0          0    0        0      0    0  
4        0        0         0          0    0        0      0    0  

[5 rows x 127 columns]

### Define A Preprocessor

Create a Scikit-Learn preprocessor that standardizes numeric variables and one-hot encodes categorical variables:

In [35]:
preprocessor = sklearn.compose.ColumnTransformer(
    transformers = [
        # Apply TfidfVectorizer to string columns
        ('title_tfidf', sklearn.feature_extraction.text.TfidfVectorizer(max_features=200, stop_words='english'), 'title'),
        ('body_tfidf', sklearn.feature_extraction.text.TfidfVectorizer(max_features=400, stop_words='english'), 'body'),
    ],
    remainder = 'drop' # Drop any other columns not explicitly listed (if any)
)

### Define A Model Pipeline: Proprocessor + Model

Combine the preprocessor with the model type in a pipeline:

In [36]:
base_classifier = sklearn.linear_model.LogisticRegression(
    random_state=42,
    solver='liblinear', # 'liblinear' is good for smaller datasets and L1/L2 regularization
    max_iter=100,      # Increase max_iter for convergence if needed
    penalty='l1',       # L1 regularization often good for sparse features
    C=0.1               # Regularization strength (tune this)
)

In [37]:
multi_label_classifier = sklearn.multioutput.MultiOutputClassifier(
    estimator=base_classifier,
    n_jobs=2 # Parallelize fitting of individual classifiers for each label
)

In [38]:
model_pipeline = sklearn.pipeline.Pipeline(
    steps = [
        ('preprocessor', preprocessor),
        ('classifier', multi_label_classifier)
    ]
)

### Train/Fit The Model

Use the training data to train the model using the pipeline:

In [39]:
X_train.shape, y_train_processed.shape

((169296, 2), (169296, 127))

In [40]:
model_pipeline.fit(X_train, y_train_processed.toarray())

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('title_tfidf',
                                                  TfidfVectorizer(max_features=200,
                                                                  stop_words='english'),
                                                  'title'),
                                                 ('body_tfidf',
                                                  TfidfVectorizer(max_features=400,
                                                                  stop_words='english'),
                                                  'body')])),
                ('classifier',
                 MultiOutputClassifier(estimator=LogisticRegression(C=0.1,
                                                                    penalty='l1',
                                                                    random_state=42,
                                                                    solver='liblinear'),
                                       n_jobs=2))])

---
## Evaluate The Model

Use the [Scikit-Learn metrics](https://scikit-learn.org/stable/api/sklearn.metrics.html) module to get evaluation metrics for the model on the data splits.

### Get Predictions

In [57]:
# 1. Get predictions from the model_pipeline
# model_pipeline.predict_proba(X_test) returns a LIST of (n_samples, 2) arrays (one array per label)
y_pred_proba_list = model_pipeline.predict_proba(X_test)
# Stack them into a single (n_samples, n_labels) array of probabilities for the positive class (1)
y_pred_proba_test = np.column_stack([p[:, 1] for p in y_pred_proba_list])

# model_pipeline.predict(X_test) returns a single (n_samples, n_labels) NumPy array
# This is ALREADY the correctly shaped binary indicator matrix.
# We just need to assign it directly and not try to `column_stack` it.
y_pred_class_test = model_pipeline.predict(X_test)

### Evaluation Metrics For Multi-Label Classification

In [59]:
# Subset Accuracy: A very strict metric for multi-label. A sample is "accurate" only if ALL its labels are correctly predicted.
# sklearn.metrics.accuracy_score can take sparse inputs for y_true and y_pred.
test_accuracy_subset = sklearn.metrics.accuracy_score(y_test_processed.toarray(), y_pred_class_test)
print(f"Subset Accuracy (TEST): {test_accuracy_subset:.4f} (all labels for a sample must be correct)")

Subset Accuracy (TEST): 0.2869 (all labels for a sample must be correct)


Subset Accuracy (TEST): 0.2869 (all labels for a sample must be correct)

This metric is very strict for multi-label classification. A score of 0.2869 means that for only about 28.7% of the samples in your test set, every single one of their associated labels was predicted correctly by the model. This is a typical value for multi-label problems, as getting all labels right for a single instance can be challenging.

In [60]:
# Micro-averaged PR AUC (Average Precision Score): Calculates PR AUC globally by summing TP, FP, FN across all labels.
# This is a common overall metric for multi-label.
test_prc_auc = sklearn.metrics.average_precision_score(y_test_processed.toarray(), y_pred_proba_test, average='micro')
print(f"Micro-averaged PR AUC (TEST): {test_prc_auc:.4f}")

Micro-averaged PR AUC (TEST): 0.5504


Micro-averaged PR AUC (TEST): 0.5504

This indicates that, globally across all labels, about 55.0% of the positive predictions made by your model were correct. This is a moderate score, suggesting room for improvement in precision.

In [61]:
# Micro-averaged ROC AUC: Calculates ROC AUC globally by summing TP, FP, FN across all labels.
# This is another common overall metric for multi-label.
test_roc_auc = sklearn.metrics.roc_auc_score(y_test_processed.toarray(), y_pred_proba_test, average='micro')
print(f"Micro-averaged ROC AUC (TEST): {test_roc_auc:.4f}")

Micro-averaged ROC AUC (TEST): 0.9508


Micro-averaged ROC AUC (TEST): 0.9508

This is a strong result. A micro-averaged ROC AUC of 0.9508 indicates that the model is performing quite well at ranking positive instances higher than negative ones, globally across all labels. It's a good sign of the model's overall discriminative power.

In [65]:
# Initialize list to store log loss for each individual label
per_label_log_losses = []

for i in range(y_test_processed.toarray().shape[1]): # Iterate through each label (column)
    y_true_label = y_test_processed.toarray()[:, i]     # True labels for the current label
    y_pred_proba_label = y_pred_proba_test[:, i]   # Predicted probabilities for the current label

    # Check if the label's true data is perfectly uniform (all 0s or all 1s)
    # log_loss might give errors or NaN if y_true_label has only one class
    if len(np.unique(y_true_label)) == 1:
        # If all true labels are the same (e.g., all 0s or all 1s), log_loss can be undefined or zero.
        # For evaluation, you might skip it or assign 0. For strictness, you could calculate.
        # However, it might be more robust to add a very small epsilon to probabilities if they are 0 or 1.
        # For simplicity, if a label has only one class, its log loss is often 0 or skipped.
        # Let's add a small epsilon to probabilities to avoid log(0) if predictions are exactly 0 or 1.
        y_pred_proba_label = np.clip(y_pred_proba_label, 1e-15, 1 - 1e-15)
        # Calculate log loss. If all true labels are the same, this might still be problematic if not handled by scikit-learn.
        # The sum will still be zero for the other class.
        pass # Let the log_loss function handle it. It should be fine as long as there are no NaNs in input

    # Calculate log loss for this binary label
    per_label_log_loss = sklearn.metrics.log_loss(y_true_label, y_pred_proba_label, labels=[0, 1]) # Specify labels to avoid warnings
    per_label_log_losses.append(per_label_log_loss)

# Sum all per-label log losses
test_log_loss_multi_label = np.sum(per_label_log_losses)

print(f"Log Loss (TEST): {test_log_loss_multi_label:.4f} (sum of per-label log-losses)")

Log Loss (TEST): 4.2172 (sum of per-label log-losses)


In [66]:
print("\nClassification Report (TEST Split - Per-Label and Averages):")
# sklearn.metrics.classification_report directly supports multi-label
# target_names provides the labels for the report rows.
# zero_division=0 prevents warnings/errors if a label has no predictions.
print(sklearn.metrics.classification_report(y_true=y_test_processed, y_pred=y_pred_class_test,
                                          target_names=mlb.classes_, zero_division=0))


Classification Report (TEST Split - Per-Label and Averages):
                            precision    recall  f1-score   support

                      .net       0.53      0.05      0.09       168
                      ajax       0.85      0.56      0.68       229
                 algorithm       0.00      0.00      0.00        93
       amazon-web-services       0.50      0.01      0.01       155
                   android       0.93      0.69      0.79      1385
            android-studio       0.59      0.16      0.25       100
                   angular       0.79      0.51      0.62       531
                 angularjs       0.86      0.53      0.65       352
                    apache       0.25      0.02      0.04        88
              apache-spark       0.82      0.10      0.18       141
                       api       0.11      0.01      0.02        79
                    arrays       0.56      0.33      0.42       426
                   asp.net       0.57      0.21      

---
## Register The Model In Vertex AI Model Registry

### Save The Model

Use the [joblib](https://joblib.readthedocs.io/en/stable/) to save the model pipeline to a local folder.

In [67]:
model_filename = os.path.join(DIR, 'model.joblib')

try:
    joblib.dump(model_pipeline, model_filename)
    print(f"Model pipeline saved successfully to: {model_filename}")
except Exception as e:
    print(f"Error saving model pipeline: {e}")

Model pipeline saved successfully to: files/evaluation-classification-multi-label/model.joblib


### Move The Model To GCS

Store the joblib file in GCS for use with the Vertex AI Model Registry.

In [68]:
blob_path = f"{SERIES}/{EXPERIMENT}/model_files"
blob_name = os.path.basename(model_filename)
blob = bucket.blob(os.path.join(blob_path, blob_name))

try:
    blob.upload_from_filename(model_filename, timeout = 600) 
    print(f"Model pipeline uploaded successfully to GCS: gs://{bucket.name}/{blob.name}")
except Exception as e:
    print(f"Error uploading model pipeline to GCS: {e}")

Model pipeline uploaded successfully to GCS: gs://statmike-mlops-349915/mlops/evaluation-classification-multi-label/model_files/model.joblib


### Register The Model

Registering the model in the [Vertex AI Model Registry](https://cloud.google.com/vertex-ai/docs/model-registry/introduction) requires:
- A GCS URI for the model files
- A Artifact Registry URI to a container than can serve the model files

Vertex AI includes prebuilt containers for [training](https://cloud.google.com/vertex-ai/docs/training/pre-built-containers) and [serving](https://cloud.google.com/vertex-ai/docs/predictions/pre-built-containers) including [these](https://cloud.google.com/vertex-ai/docs/predictions/pre-built-containers#scikit-learn) that are purpose built for Scikit-Learn.  You can alos customize these containers (or your own container) as a [custom prediction container](https://cloud.google.com/vertex-ai/docs/predictions/use-custom-container).  For an example check out this workflow in this repository: [Catboost Custom Prediction With FastAPI](../../Framework%20Workflows/CatBoost/CatBoost%20Custom%20Prediction%20With%20FastAPI.ipynb).

Check for existing version of the model:

In [69]:
parent_model = next(
    (model for model in aiplatform.Model.list(filter=f'display_name="{SERIES}-{EXPERIMENT}"')),
    None # Default value if no model/version is found
)
parent_model

Upload the model to the registry as a new version which will be version 1 if no prior version was found:

In [70]:
vertex_model = aiplatform.Model.upload(
    display_name = SERIES+'-'+EXPERIMENT,
    model_id = SERIES+'-'+EXPERIMENT,
    parent_model = parent_model,
    serving_container_image_uri = 'us-docker.pkg.dev/vertex-ai/prediction/sklearn-cpu.1-5:latest',
    artifact_uri = f'gs://{bucket.name}/{blob_path}',
    is_default_version = True,
    labels = {'series': SERIES, 'experiment': EXPERIMENT}
)

Creating Model
Create Model backing LRO: projects/1026793852137/locations/us-central1/models/mlops-evaluation-classification-multi-label/operations/8446834631113703424
Model created. Resource name: projects/1026793852137/locations/us-central1/models/mlops-evaluation-classification-multi-label@1
To use this Model in another session:
model = aiplatform.Model('projects/1026793852137/locations/us-central1/models/mlops-evaluation-classification-multi-label@1')


---
## Load Model Evaluations To The Vertex AI Model Registry

The versioned models in Vertex AI Model Registry can include evaluations.  Evaluation can even be vizualized and compared directly in the console.  

The model registry, and Vertex AI SDK, include model evaluations as a service where managed jobs can be used to calculate predictions and automatically laod evaluation metrics based on the model type (regression, classification, or forecasting). Read more about [Model evaluation in Vertex AI](https://cloud.google.com/vertex-ai/docs/evaluation/introduction).

In this workflow the evaluation will be manually uploaded with the Vertex AI SDK after first being prepared locally.  The evaluation need to be prepared in the format of one of the schems provided [here](https://console.cloud.google.com/storage/browser/google-cloud-aiplatform/schema/modelevaluation).

This part of the workflow will demonstrate preparing the full set of evaluation metrics for the model above on each split of the data and then loading in the versioned model in the model registry.  The evaluation will then be shown in the console, and retrieved with the SDK.


### Structured Evaluation

First, prepare the evaluations for the model in the structure of the provided schema `google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml` found in the GCS location that is publically available here: [google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml](https://console.cloud.google.com/storage/browser/google-cloud-aiplatform/schema/modelevaluation).

In this example the complete schema is used but it is possible to provide a limited subset.

In [80]:
class_labels = mlb.classes_.tolist()
n_classes = len(class_labels)

evaluations = {}

data_splits = {
    'TRAIN': (X_train, y_train_processed),
    'VALIDATE': (X_val, y_val_processed),
    'TEST': (X_test, y_test_processed)
}

In [81]:
for split_name, (X_split, y_split_sparse) in data_splits.items():   
    # Convert y_split_sparse (true labels) to dense for metrics compatibility and easier flattening
    y_split_dense = y_split_sparse.toarray()
    # predictions
    y_pred_proba_list = model_pipeline.predict_proba(X_split)
    y_pred_proba = np.column_stack([p[:, 1] for p in y_pred_proba_list])
    y_pred_class = model_pipeline.predict(X_split)

    # --- Prepare for overall binary confusion matrix (micro-averaged) ---
    # Flatten true and predicted matrices for micro-averaging across all classes for overall confusion matrix
    y_true_flat_overall = y_split_dense.ravel()
    # For overall confusion matrix, use y_pred_class (the 0/1 predictions)
    y_pred_flat_overall = y_pred_class.ravel()
    tn_flat_overall, fp_flat_overall, fn_flat_overall, tp_flat_overall = sklearn.metrics.confusion_matrix(y_true_flat_overall, y_pred_flat_overall, labels=[0, 1]).ravel()


    # overall metrics (micro-averaged for multi-label schema compliance)
    au_prc_val = sklearn.metrics.average_precision_score(y_split_dense, y_pred_proba, average='micro')
    au_roc_val = sklearn.metrics.roc_auc_score(y_split_dense, y_pred_proba, average='micro')
    
    # --- Update logloss to the loop that calculates the sum of logloss over all labels ---
    per_label_log_losses = []
    for i in range(y_split_dense.shape[1]): # Iterate through each label
        y_true_label = y_split_dense[:, i]
        y_pred_proba_label = y_pred_proba[:, i]
        # Clip probabilities to avoid log(0) issues for log_loss
        y_pred_proba_label_clipped = np.clip(y_pred_proba_label, 1e-15, 1 - 1e-15)
        # Calculate log loss for this binary label
        per_label_log_losses.append(sklearn.metrics.log_loss(y_true_label, y_pred_proba_label_clipped, labels=[0, 1]))
    log_loss_val = np.sum(per_label_log_losses)
    # --- End logloss update ---

    # Overall Confusion Matrix: Now adding the micro-averaged binary confusion matrix
    # No confusionMatrix key is added to the top-level evaluation dictionary for this phase.
    # We will add it to the final dict structure.
    overall_binary_conf_matrix = {
        'annotationSpecs': [{'id': '0', 'displayName': 'Negative'}, {'id': '1', 'displayName': 'Positive'}],
        'rows': [{'dataItemCounts': sklearn.metrics.confusion_matrix(y_true_flat_overall, y_pred_flat_overall, labels=[0, 1])[0].tolist()},
                 {'dataItemCounts': sklearn.metrics.confusion_matrix(y_true_flat_overall, y_pred_flat_overall, labels=[0, 1])[1].tolist()}]
    }

    # confidence metrics for splits
    confidence_thresholds = (
        list(np.arange(0.0, 1.0, 0.05)) +
        list(np.arange(0.95, 1.0, 0.01))
    )
    confidence_thresholds = sorted(list(set([round(t, 2) for t in confidence_thresholds])))
    confidence_metrics_list = []
    for threshold in confidence_thresholds:
        # Apply threshold to probabilities for each class independently to get a binary matrix
        y_pred_binary_matrix_at_threshold = (y_pred_proba >= threshold).astype(int)
        
        # Flatten true and predicted matrices for micro-averaging across all classes
        y_true_flat = y_split_dense.ravel()
        y_pred_flat = y_pred_binary_matrix_at_threshold.ravel()

        # Calculate micro-averaged counts (TP, FP, FN, TN) from flattened data
        # sklearn.metrics.confusion_matrix on flattened arrays is robust for this
        tn_flat, fp_flat, fn_flat, tp_flat = sklearn.metrics.confusion_matrix(y_true_flat, y_pred_flat, labels=[0, 1]).ravel()
        
        # Calculate micro-averaged metrics from these flattened counts, handling division by zero
        recall_val = tp_flat / (tp_flat + fn_flat) if (tp_flat + fn_flat) > 0 else 0.0
        precision_val = tp_flat / (tp_flat + fp_flat) if (tp_flat + fp_flat) > 0 else 0.0
        f1_val = 2 * (precision_val * recall_val) / (precision_val + recall_val) if (precision_val + recall_val) > 0 else 0.0
        fpr_val = fp_flat / (fp_flat + tn_flat) if (fp_flat + tn_flat) > 0 else 0.0

        confidence_metrics_entry = {
            'confidenceThreshold': threshold,
            # 'maxPredictions': INT32_MAX_VALUE, # User comments this out intentionally
            'recall': recall_val,
            'precision': precision_val,
            'falsePositiveRate': fpr_val,
            'f1Score': f1_val,
            'truePositiveCount': int(tp_flat),
            'falsePositiveCount': int(fp_flat),
            'falseNegativeCount': int(fn_flat),
            'trueNegativeCount': int(tn_flat),
            # --- Adding binary confusion matrix for each confidence threshold ---
            'confusionMatrix': {
                'annotationSpecs': [{'id': '0', 'displayName': 'Negative'}, {'id': '1', 'displayName': 'Positive'}],
                'rows': [{'dataItemCounts': sklearn.metrics.confusion_matrix(y_true_flat, y_pred_flat, labels=[0, 1])[0].tolist()},
                         {'dataItemCounts': sklearn.metrics.confusion_matrix(y_true_flat, y_pred_flat, labels=[0, 1])[1].tolist()}]
            }
            # Metrics like precisionAt1, recallAt1 etc. are not standard for multi-label thresholding and are omitted.
        }
        
        confidence_metrics_list.append(confidence_metrics_entry)
        
    # Populate the main evaluations dictionary for the current split
    evaluations[split_name] = {}
    evaluations[split_name]['metrics'] = {
        'auPrc': au_prc_val,
        'auRoc': au_roc_val,
        'logLoss': log_loss_val,
        'confidenceMetrics': confidence_metrics_list,
        # --- Adding overall binary confusion matrix ---
        'confusionMatrix': overall_binary_conf_matrix
    }

Review the metrics for the first threshold of the 'TRAIN' split:

In [82]:
evaluations['TRAIN']['metrics']['confidenceMetrics'][0]

{'confidenceThreshold': np.float64(0.0),
 'recall': np.float64(1.0),
 'precision': np.float64(0.0121506886880138),
 'falsePositiveRate': np.float64(1.0),
 'f1Score': np.float64(0.024009643670279888),
 'truePositiveCount': 261247,
 'falsePositiveCount': 21239345,
 'falseNegativeCount': 0,
 'trueNegativeCount': 0,
 'confusionMatrix': {'annotationSpecs': [{'id': '0',
    'displayName': 'Negative'},
   {'id': '1', 'displayName': 'Positive'}],
  'rows': [{'dataItemCounts': [0, 21239345]},
   {'dataItemCounts': [0, 261247]}]}}

In [83]:
#evaluations['TRAIN']['metrics']

### Load The Evaluations

Use the [ModelServiceClient](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service) to load the evaluation metrics with the [import_model_evaluation](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service.ModelServiceClient#google_cloud_aiplatform_v1_services_model_service_ModelServiceClient_import_model_evaluation) method.

In [84]:
model_client = aiplatform.gapic.ModelServiceClient(
    client_options = {'api_endpoint': f"{REGION}-aiplatform.googleapis.com"}
)

In [85]:
for eval in evaluations:
    eval_upload = model_client.import_model_evaluation(
        parent = vertex_model.versioned_resource_name,
        model_evaluation = aiplatform.gapic.ModelEvaluation(
            display_name = f'training-{eval.lower()}',
            metrics_schema_uri = 'gs://google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml',
            metrics = evaluations[eval]['metrics']
        )
    )
    evaluations[eval]['vertex_resource'] = {'name': eval_upload.name, 'display_name': eval_upload.display_name}

In [86]:
evaluations['TRAIN']['vertex_resource']

{'name': 'projects/1026793852137/locations/us-central1/models/7126562680755716096/evaluations/2361194751273445101',
 'display_name': 'training-train'}

### Review Evaluations In The Vertex AI Console

See a list of evaluations for a model+version in the Vertex AI Model Registry:
<div style="text-align: center;"><img src="../resources/images/screenshots/evaluation/class-multi-label/registry_evaluations.png" style="border: 2px solid black; width: auto; height: 600px;"></div>

Review an evaluation, including changing the confidence threshold:
<div style="text-align: center;"><img src="../resources/images/screenshots/evaluation/class-multi-label/registry_evaluation_review.gif" style="border: 2px solid black; width: auto; height: 600px;"></div>

Compare evaluations side-by-side:
<div style="text-align: center;"><img src="../resources/images/screenshots/evaluation/class-multi-label/registry_evaluation_compare.gif" style="border: 2px solid black; width: auto; height: 600px;"></div>


In [87]:
vertex_model.get_model_evaluation(evaluation_id = eval_upload.name.split('/')[-1])

resource name: projects/1026793852137/locations/us-central1/models/mlops-evaluation-classification-multi-label@1/evaluations/7779771552308240837

### Retrieve The Model Evaluations: Vertex AI SDK

Use the same model object we created above to laod the model to the registry to also retrieve the evaluations.
- [Model.list_model_evaluations](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model#google_cloud_aiplatform_Model_list_model_evaluations)

In [88]:
vertex_eval_list = vertex_model.list_model_evaluations()
[e.display_name for e in vertex_eval_list]

['training-validate', 'training-train', 'training-test']

In [89]:
dict(vertex_eval_list[0].metrics)

{'auPrc': 0.5535403,
 'auRoc': 0.9511465,
 'logLoss': 4.194654,
 'confusionMatrix': <proto.marshal.collections.maps.MapComposite at 0x7f3e9a9aaae0>,
 'confidenceMetrics': [<proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9ab680>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9a9340>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9aac90>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9a9b20>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9a98b0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9a9760>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9a8bf0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9aaea0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9abbf0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9abb60>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9ab500>, <proto.marshal.collections.maps.MapC

In [90]:
dict(vertex_eval_list[0].metrics['confusionMatrix'])

{'rows': [[2651564.0, 3505.0], [20940.0, 11692.0]],
 'annotationSpecs': [<proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9aa270>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9ab620>]}

In [91]:
type(vertex_eval_list[0])

google.cloud.aiplatform.model_evaluation.model_evaluation.ModelEvaluation

In [92]:
vertex_eval_list[0].to_dict()

{'name': 'projects/1026793852137/locations/us-central1/models/mlops-evaluation-classification-multi-label@1/evaluations/1230998364301100316',
 'metricsSchemaUri': 'gs://google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml',
 'metrics': {'auPrc': 0.5535403,
  'auRoc': 0.9511465,
  'logLoss': 4.194654,
  'confusionMatrix': {'rows': [[2651564.0, 3505.0], [20940.0, 11692.0]],
   'annotationSpecs': [{'displayName': 'Negative', 'id': '0'},
    {'displayName': 'Positive', 'id': '1'}]},
  'confidenceMetrics': [{'precision': 0.012141231,
    'falseNegativeCount': '0',
    'truePositiveCount': '32632',
    'recall': 1.0,
    'trueNegativeCount': '0',
    'confidenceThreshold': 0.0,
    'falsePositiveRate': 1.0,
    'falsePositiveCount': '2655069',
    'confusionMatrix': {'rows': [[0.0, 2655069.0], [0.0, 32632.0]],
     'annotationSpecs': [{'displayName': 'Negative', 'id': '0'},
      {'displayName': 'Positive', 'id': '1'}]},
    'f1Score': 0.02399118},
   {'precision'

### Retrieve The Model Evaluations: Vertex AI SDK ModelServiceClient

Use the [ModelServiceClient](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service) to retreive the evaluation metrics with the [list_model_evaluations](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service.ModelServiceClient#google_cloud_aiplatform_v1_services_model_service_ModelServiceClient_list_model_evaluations) method.

In [93]:
eval = model_client.get_model_evaluation(name = evaluations['TRAIN']['vertex_resource']['name'])

In [94]:
dict(eval.metrics)

{'auPrc': 0.55779755,
 'auRoc': 0.95273966,
 'logLoss': 4.166204,
 'confusionMatrix': <proto.marshal.collections.maps.MapComposite at 0x7f3e9a9d17f0>,
 'confidenceMetrics': [<proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1190>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1130>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1010>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d0fb0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1250>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d12e0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1430>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d18b0>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1910>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d1970>, <proto.marshal.collections.maps.MapComposite object at 0x7f3e9a9d19d0>, <proto.marshal.collections.maps.Ma

In [95]:
type(eval)

google.cloud.aiplatform_v1.types.model_evaluation.ModelEvaluation

## Sliced Evaluations

For any model it can be beneficial to review metrics on slices of data.  For a multi-label classifier like this it is directly helpful to review the metrics related to each possible classification label as slices.  In this section we create evaluation metrics for the binary case of each possible label and load them as slices for review along with the overall evaluation metrics above.

### Structured Evaluation For Slices

Similar to above, prepare the evaluations for the model in the structure of the provided schema `google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml` found in the GCS location that is publically available here: [google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml](https://console.cloud.google.com/storage/browser/google-cloud-aiplatform/schema/modelevaluation).

The difference here is doing these structured evalautions on the label, each class, separately as a binary comparision: not in class, is in class.

In this example the complete schema is used but it is possible to provide a limited subset.

In [96]:
# Main loop over each data split
for split_name, (X_split, y_split_sparse) in data_splits.items():
    
    # Convert y_split_sparse (true labels) to dense for metrics compatibility and easier flattening
    y_split_dense = y_split_sparse.toarray()

    # Perform predictions once per split
    # For MultiOutputClassifier, predict_proba returns a list of (n_samples, 2) arrays.
    y_pred_proba_list = model_pipeline.predict_proba(X_split)
    # Stack them into a single (n_samples, n_labels) array of probabilities for the positive class (1)
    y_pred_proba = np.column_stack([p[:, 1] for p in y_pred_proba_list])

    # Loop over each label (tag) to treat it as a one-vs-rest "slice"
    evaluations[split_name]['slices'] = {} # Initialize 'slices' dictionary for this split
    for slice_index, slice_value in enumerate(class_labels): # Iterate using index and actual label name
        
        # --- Start of Integrated Binary Metrics Block for Slices ---

        # 1. Binarize the problem for the current slice (extract column for this label)
        y_true_binary = y_split_dense[:, slice_index] # True labels for this specific tag
        y_pred_proba_slice = y_pred_proba[:, slice_index] # Predicted probabilities for this specific tag
        
        # --- Add check for sufficient data in this slice ---
        # Skip evaluation for slices (labels) with too few positive samples or if it's a single-class slice.
        # This prevents errors in metrics like ROC AUC, LogLoss, and improves stability.
        # Heuristic: e.g., at least 20 positive samples needed for meaningful metrics.
        if y_true_binary.sum() < 20 or len(np.unique(y_true_binary)) < 2:
            # print(f"    Skipping slice '{slice_value}' in {split_name} due to insufficient positive samples or single class (count: {y_true_binary.sum()}).")
            continue # Skip this slice if not enough data

        # 2. Calculate overall slice metrics using your binary logic
        y_pred_class_default_binary = (y_pred_proba_slice >= 0.5).astype(int)
        
        au_prc_slice = sklearn.metrics.average_precision_score(y_true_binary, y_pred_proba_slice)
        au_roc_slice = sklearn.metrics.roc_auc_score(y_true_binary, y_pred_proba_slice)
        
        # Corrected: Log loss for a single binary label (clip probabilities to avoid log(0))
        y_pred_proba_slice_clipped = np.clip(y_pred_proba_slice, 1e-15, 1 - 1e-15)
        log_loss_slice = sklearn.metrics.log_loss(y_true_binary, y_pred_proba_slice_clipped, labels=[0, 1])
        
        overall_conf_matrix_slice = sklearn.metrics.confusion_matrix(y_true_binary, y_pred_class_default_binary, labels=[0, 1])
        
        # Create dynamic annotations for the 2x2 confusion matrix for this slice
        slice_annotations = [
            {'id': '0', 'displayName': f'Not {slice_value}'}, # Negative class (0)
            {'id': '1', 'displayName': slice_value}          # Positive class (1)
        ]

        # 3. Calculate confidence metrics using your binary thresholding logic
        confidence_thresholds = sorted(list(set([round(t, 2) for t in list(np.arange(0.0, 1.0, 0.05)) + list(np.arange(0.95, 1.01, 0.01))])))
        confidence_metrics_list_slice = []
        for threshold in confidence_thresholds:
            y_pred_at_threshold = (y_pred_proba_slice >= threshold).astype(int)
            
            cm_at_threshold = sklearn.metrics.confusion_matrix(y_true_binary, y_pred_at_threshold, labels=[0, 1])
            tn, fp, fn, tp = cm_at_threshold.ravel()

            # Calculate metrics
            recall_val = sklearn.metrics.recall_score(y_true_binary, y_pred_at_threshold, zero_division=0)
            precision_val = sklearn.metrics.precision_score(y_true_binary, y_pred_at_threshold, zero_division=0)
            false_positive_rate_val = fp / (fp + tn) if (fp + tn) > 0 else 0.0
            f1_score_val = sklearn.metrics.f1_score(y_true_binary, y_pred_at_threshold, zero_division=0)

            # Assemble the dictionary for this threshold exactly as you specified
            confidence_metrics_entry = {
                'confidenceThreshold': threshold,
                # 'maxPredictions': INT32_MAX_VALUE, # User comments this out intentionally
                'recall': recall_val,
                'precision': precision_val,
                'falsePositiveRate': false_positive_rate_val,
                'f1Score': f1_score_val,
                'truePositiveCount': int(tp),
                'falsePositiveCount': int(fp),
                'falseNegativeCount': int(fn),
                'trueNegativeCount': int(tn),
                'confusionMatrix': { # Optional confusion matrix for each threshold (2x2 for binary slice)
                    'annotationSpecs': slice_annotations, # Use specific slice annotations (Not/Tag_Name)
                    'rows': [{'dataItemCounts': row.tolist()} for row in cm_at_threshold] # Use dataItemCounts
                }
            }
            # Remove 'At1' metrics (precisionAt1, recallAt1, f1ScoreAt1, falsePositiveRateAt1)
            # They are not included as per previous decisions for multi-label overall metrics.
            
            confidence_metrics_list_slice.append(confidence_metrics_entry)

        # 4. Assemble the final dictionary for the entire slice
        evaluations[split_name]['slices'][slice_value] = {
            'auPrc': au_prc_slice,
            'auRoc': au_roc_slice,
            'logLoss': log_loss_slice,
            'confidenceMetrics': confidence_metrics_list_slice,
            'confusionMatrix': { # Overall confusion matrix for this slice (2x2 for binary slice)
                'annotationSpecs': slice_annotations, # Use specific slice annotations (Not/Tag_Name)
                'rows': [{'dataItemCounts': row.tolist()} for row in overall_conf_matrix_slice] # Use dataItemCounts
            }
        }
        # === End of Integrated Block ===

In [97]:
evaluations['TRAIN']['slices'].keys()

dict_keys(['.net', 'ajax', 'algorithm', 'amazon-web-services', 'android', 'android-studio', 'angular', 'angularjs', 'apache', 'apache-spark', 'api', 'arrays', 'asp.net', 'asp.net-core', 'asp.net-mvc', 'authentication', 'azure', 'bash', 'c', 'c#', 'c++', 'c++11', 'codeigniter', 'cordova', 'css', 'csv', 'database', 'dataframe', 'date', 'datetime', 'dictionary', 'django', 'docker', 'eclipse', 'elasticsearch', 'entity-framework', 'excel', 'express', 'file', 'firebase', 'firebase-realtime-database', 'for-loop', 'forms', 'function', 'git', 'go', 'google-chrome', 'google-maps', 'hibernate', 'html', 'image', 'ionic-framework', 'ionic2', 'ios', 'java', 'javascript', 'jenkins', 'jquery', 'json', 'laravel', 'laravel-5', 'linq', 'linux', 'list', 'loops', 'machine-learning', 'macos', 'matlab', 'matplotlib', 'maven', 'mongodb', 'multithreading', 'mysql', 'node.js', 'numpy', 'objective-c', 'opencv', 'oracle', 'pandas', 'performance', 'php', 'postgresql', 'powershell', 'python', 'python-2.7', 'python-

In [98]:
evaluations['TRAIN']['slices']['.net']['confidenceMetrics'][0]

{'confidenceThreshold': np.float64(0.0),
 'recall': 1.0,
 'precision': 0.008381769208959456,
 'falsePositiveRate': np.float64(1.0),
 'f1Score': 0.01662419822511203,
 'truePositiveCount': 1419,
 'falsePositiveCount': 167877,
 'falseNegativeCount': 0,
 'trueNegativeCount': 0,
 'confusionMatrix': {'annotationSpecs': [{'id': '0',
    'displayName': 'Not .net'},
   {'id': '1', 'displayName': '.net'}],
  'rows': [{'dataItemCounts': [0, 167877]}, {'dataItemCounts': [0, 1419]}]}}

### Load The Evaluation Slices To The Existing Evaluation

The evaluations slices are added to the evaluations already created and loaded above.  To do this use the [ModelServiceClient](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service) to load the sliced evaluation metrics with the [batch_import_model_evaluation_slices](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service.ModelServiceClient#google_cloud_aiplatform_v1_services_model_service_ModelServiceClient_batch_import_model_evaluation_slices) method.

**NOTE:** This load the first 50 labels as slices to the evaluations created above.

In [102]:
for eval in evaluations:

    eval_upload = model_client.batch_import_model_evaluation_slices(
        parent = evaluations[eval]['vertex_resource']['name'],
        model_evaluation_slices = [
            aiplatform.gapic.ModelEvaluationSlice(
                slice_ = {'dimension': 'annotationSpec', 'value': slice},
                metrics_schema_uri = 'gs://google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml',
                metrics = evaluations[eval]['slices'][slice]
            )
            for s, slice in enumerate(evaluations[eval]['slices']) if s < 50
        ]
    )

### Review Evaluations **With Slices** In The Vertex AI Console

Review an evaluation, including changing the individual slices:
<div style="text-align: center;"><img src="../resources/images/screenshots/evaluation/class-multi-label/registry_evaluation_slices.gif" style="border: 2px solid black; width: auto; height: 600px;"></div>

### Retrieve The Model Evaluation **Slices**: Vertex AI SDK ModelServiceClient

Use the [ModelServiceClient](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service) to retreive the evaluation metrics with the [list_model_evaluations](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1.services.model_service.ModelServiceClient#google_cloud_aiplatform_v1_services_model_service_ModelServiceClient_list_model_evaluations) method.

In [105]:
evaluations[eval]['vertex_resource']['name']

'projects/1026793852137/locations/us-central1/models/7126562680755716096/evaluations/7779771552308240837'

In [106]:
eval_slices = model_client.list_model_evaluation_slices(parent = evaluations[eval]['vertex_resource']['name'])
type(eval_slices)

google.cloud.aiplatform_v1.services.model_service.pagers.ListModelEvaluationSlicesPager

In [107]:
eval_slice = next(iter(eval_slices))
type(eval_slice)

google.cloud.aiplatform_v1.types.model_evaluation_slice.ModelEvaluationSlice

In [108]:
eval_slice

name: "projects/1026793852137/locations/us-central1/models/7126562680755716096/evaluations/7779771552308240837/slices/272721240179179034"
slice_ {
  dimension: "annotationSpec"
  value: "date"
}
metrics_schema_uri: "gs://google-cloud-aiplatform/schema/modelevaluation/classification_metrics_1.0.0.yaml"
metrics {
  struct_value {
    fields {
      key: "logLoss"
      value {
        number_value: 0.015195765
      }
    }
    fields {
      key: "confusionMatrix"
      value {
        struct_value {
          fields {
            key: "rows"
            value {
              list_value {
                values {
                  list_value {
                    values {
                      number_value: 21066
                    }
                    values {
                      number_value: 10
                    }
                  }
                }
                values {
                  list_value {
                    values {
                      number_value: 77
    

---
## More Ways To Work With Evaluations

The [Vertex AI Experiment service](https://cloud.google.com/vertex-ai/docs/experiments/intro-vertex-ai-experiments) is another excellent way to store and track experiments as well as other model information.  Learn more in this repositiory [starting here](../Experiment%20Tracking/readme.md).